packages/flutter/lib/src/rendering/object.dart

```diff
@@ -3248,31 +3260,64 @@ abstract class RenderObject with DiagnosticableTreeMixin implements HitTestTarge
   /// use [applyPaintTransform] from the [RenderView] to further transform the
   /// coordinate.
-  Matrix4 getTransformTo(RenderObject? ancestor) {
-    final bool ancestorSpecified = ancestor != null;
+  Matrix4 getTransformTo(RenderObject? target, { bool skipIfInUnpaintedSubtree = false }) {
     assert(attached);
-    if (ancestor == null) {
-      final RenderObject? rootNode = owner!.rootNode;
-      if (rootNode is RenderObject) {
-        ancestor = rootNode;
+    // The paths to fromRenderObject and toRenderObject's common ancestor.
+    List<RenderObject>? fromPath;
+    List<RenderObject>? toPath;
+
+    RenderObject from = this;
+    RenderObject to = target ?? owner!.rootNode!;
+
+    while (!identical(from, to)) {
+      final int fromDepth = from.depth;
+      final int toDepth = to.depth;
+
+      if (fromDepth >= toDepth) {
+        assert(from.parent != null, '$target and $this are not in the same render tree.');
+        final RenderObject fromParent = from.parent!;
+        if (skipIfInUnpaintedSubtree && !fromParent.paintsChild(from)) {
+          return Matrix4.zero();
+        }
+        (fromPath ??= <RenderObject>[this]).add(fromParent);
+        from = fromParent;
+      }
+      if (fromDepth <= toDepth) {
+        assert(to.parent != null, '$target and $this are not in the same render tree.');
+        assert(target != null);
+        final RenderObject toParent = to.parent!;
+        if (skipIfInUnpaintedSubtree && !toParent.paintsChild(to)) {
+          return Matrix4.zero();
+        }
+        (toPath ??= target == null ? <RenderObject>[] : <RenderObject>[target]).add(toParent);
+        to = toParent;
       }
     }
-    final List<RenderObject> renderers = <RenderObject>[];
-    for (RenderObject renderer = this; renderer != ancestor; renderer = renderer.parent!) {
-      renderers.add(renderer);
-      assert(renderer.parent != null); // Failed to find ancestor in parent chain.
+
+    Matrix4? fromTransform;
+    if (fromPath != null) {
+      assert(fromPath.length > 1);
+      fromTransform = Matrix4.identity();
+      final int lastIndex = target == null ? fromPath.length - 2 : fromPath.length - 1;
+      for (int index = lastIndex; index > 0; index -= 1) {
+        fromPath[index].applyPaintTransform(fromPath[index - 1], fromTransform);
+      }
     }
-    if (ancestorSpecified) {
-      renderers.add(ancestor!);
+    if (toPath == null) {
+      return fromTransform ?? Matrix4.identity();
     }
-    final Matrix4 transform = Matrix4.identity();
-    for (int index = renderers.length - 1; index > 0; index -= 1) {
-      renderers[index].applyPaintTransform(renderers[index - 1], transform);
+
+    assert(toPath.length > 1);
+    final Matrix4 toTransform = Matrix4.identity();
+    for (int index = toPath.length - 1; index > 0; index -= 1) {
+      toPath[index].applyPaintTransform(toPath[index - 1], toTransform);
     }
-    return transform;
+    if (toTransform.invert() == 0) {
```

Comment:
Maybe I am a bit rusty on my matrix math, but why is it for the operation below important that it is invertible? Maybe leave a comment about that?

Reply:
The implementation does nothing (https://master-api.flutter.dev/flutter/vector_math_64/Matrix4/copyInverse.html) when the matrix is singular. However I think that's because of the limitation of `applyPaintTransform`.